docs/upgrade/4.x.x.adoc

```diff
@@ -0,0 +1,475 @@
+has been moved from lib-react4xp to @enonic/react4xp.
+
+=== Content Studio
+
+SSR without hydration used to be enforced for both `edit` and `inline` mode in Content Studio.
+
+Now React4xp will only enforce SSR without hydration in Content Studio `edit` mode. 
+
+`inline` mode will use normal React4xp rendering, just as `preview` and `live` mode.
+
+=== Gradle
+
+We have simplified the gradle setup a lot.
+
+All old references to React4xp in the build.gradle file, must be removed.
+
+
+These files no longer exist:
+
+* node_modules/@enonic/react4xp/react4xp.gradle
+* node_modules/@enonic/react4xp/npmInstall.gradle
+* node_modules/@enonic/react4xp/updaters.gradle
+
+See more under the <<_build_gradle, build.gradle>> section below.
+
+=== Enonic XP dev mode
+
+When running Enonic XP in dev mode, it may be faster to build without using gradle at all.
+
+See the required changes to the build.gradle and package.json files in the How section below.
+
+=== System environment variables
+
+When building with gradle, it will automatically set some system environment variables for you.
+
+However if you want to build without using gradle you have to set them up on your own.
+
+These two are required:
+
+* R4X_APP_NAME (find the appName in gradle.properties)
+* R4X_DIR_PATH_ABSOLUTE_PROJECT (cwd/pwd)
+
+These two are optional:
+
+* R4X_BUILD_LOG_LEVEL (use INFO to get some extra logging when building)
+* NODE_ENV (the default is production, set it to development for no hashing, nor minification, etc...)
+
+=== Component props (react4xpId)
+
+React4xp used to add an extra prop called react4xpId, which was used during
+clientSide hydration and rendering. This prop is no longer needed as it is
+provided by other means (`script[data-react4xp-ref]`).
+
+So now: React component props are just normal React component props :)
+
+=== document polyfill
+
+Several frameworks and node modules uses `document` to determine whether it's code
+is running on the server, or in the browser. By polyfilling `document` that
+logic is broken. So React4xp is no longer polyfilling `document`.
+
+TIP: If your code is using `document`, and is not a pure client-side component:
+you should wrap the code with an if block to avoid that code being run on the
+server.
+
+WARNING: If you are importing some "broken" module that uses `document` without
+checking for client or server, you may polyfill `document` on your own, but it
+might break other modules which now thinks the server is the client...
+
+== How
+
+=== build.gradle
+
+```build.gradle
+dependencies {
+    include "com.enonic.lib:lib-react4xp:4.x.x"
+}
+```
+
+Remove all the old react4xp* tasks from your build.gradle file.
+
+If your project is based on an earlier version of the starter-react4xp also remove the reac4xp plugin
+
+```build.gradle 
+plugins {
+  id 'react4xp' // Delete this line
+}
+```
+
+You can probably also delete the entire buildSrc folder from your project.
```

Comment:
I'd place this as the last sentence on this section since it breaks the reasoning about changes to build.gradle

Reply:
agreed, fixed